qt: Fix font size and scaling issues

xdustinface opened the pull request:
This PR should fix all remaining places where the font scale wasn't applied yet (besides the splashscreen which probably also should follow anytime later). It also in general improves the font-scaling process and from now on it only updates fonts where really required. See commit messages. 

@thephez This should fix all the issues you pointed out in slack
@PastaPastaPasta 64ce1b5 might fix the one you sent me. At least it makes it better here on mac where i had it similar like you with a low font scale though.  
@strophy 82c1612 could be a fix for your "I can't get the same menu size like in other applications" thing

PastaPastaPasta commented:
Stuff still getting cropped in Network traffic for me

![cropped](https://user-images.githubusercontent.com/6443210/94296758-e0ac2280-ff52-11ea-823c-730478ca3f68.png)

xdustinface commented:
@PastaPastaPasta Please look at the network traffic stats again with 3c52ea3

@PastaPastaPasta @UdjinM6 I rebased this PR now on #3717 to include 0156d4d in here as it seems to make more sense in here. Should fix the min width issue. Thoughts?

xdustinface commented:
Okaay one more.. see the last three commits i changed it to resize depending on the font size. Now if any of the appearance attributes gets changed or if the PS UI gets shown/hidden the OptionsDialog gets resized to the required size.

UdjinM6 commented:
Seems to be working as expected for me (could use `QTimer::singleShot(0, this, SLOT(updateWidth()));` instead of calling it directly in ctor to fix issues with buttons visibility not being available at that point but 3170db0 is fine too :) )

xdustinface commented:
> Seems to be working as expected for me (could use `QTimer::singleShot(0, this, SLOT(updateWidth()));` instead of calling it directly in ctor to fix issues with buttons visibility not being available at that point but [3170db0](https://github.com/dashpay/dash/commit/3170db0c67247d29f6ba6f9be1d560dfe1f49596) is ~fine~ ok-ish too :) ).

Does the single shot timer 100% guaranteed only fire after the widgets visibility state flipped? I have no idea just curious. 

> EDIT: Technically speaking though, you could one day want a visible but disabled tab there and [3170db0](https://github.com/dashpay/dash/commit/3170db0c67247d29f6ba6f9be1d560dfe1f49596) won't work.

Dropped 3170db0c67247d29f6ba6f9be1d560dfe1f49596 and added fe3ae7e that should do it better

UdjinM6 commented on src/qt/optionsdialog.cpp:
```diff
@@ -459,6 +458,13 @@ void OptionsDialog::updateWidth()
+    if (!event->spontaneous()) {
+        updateWidth();
+    }
+}
```
```suggestion
    QDialog::showEvent(event);
}
```

UdjinM6 commented on src/qt/optionsdialog.h:
```diff
@@ -77,6 +77,8 @@ private Q_SLOTS:
     AppearanceWidget* appearance;
     bool fPrivateSendEnabledPrev{false};
+
+    void showEvent(QShowEvent* event);
```
```suggestion
    void showEvent(QShowEvent* event) override;
```

UdjinM6 reviewed (commented):
>Does the single shot timer 100% guaranteed only fire after the widgets visibility state flipped?

Yes, this timer event should be the last one in the event queue afaik

> Dropped 3170db0 and added 75f4389 that should do it better

Mostly works for me :) two suggestions, see below

UdjinM6 commented:
Working as expected for me now (but I had no issues with Network traffic before, so can't say if it was also improved or not).

xdustinface commented:
> Working as expected for me now (but I had no issues with Network traffic before, so can't say if it was also improved or not).

Yeah i also had no issues with it since it was not looking bad on macOS but now it should just scale as expected based on the really required size not on the "on macOS required" size :D 

Added three more commits: 

d94655e + 62a6cbc To give the main window the same resizing behaviour depending on the enabled tabs like in the options window. For me on my system the 980 min width is reasonable but not sure if other systems require more there? @PastaPastaPasta how is min size looking there if PS + Masternode are hidden? 

35dc39b To allow manual resizing after the window has been resized dynamically after restart/reopen or appearance adjustments.

PastaPastaPasta commented:
Below are the only few remaining problems I could find on really edge case font settings. Don't think it's critical to get these fixed


![Screenshot from 2020-09-26 15-31-18](https://user-images.githubusercontent.com/6443210/94348837-e9722680-002e-11eb-8e5f-256e45ee91e6.png)

###########################################

![Screenshot from 2020-09-26 15-28-33](https://user-images.githubusercontent.com/6443210/94348838-e9722680-002e-11eb-95ed-6bdcf6e007d4.png)

###########################################

![Screenshot from 2020-09-26 15-27-59](https://user-images.githubusercontent.com/6443210/94348839-ea0abd00-002e-11eb-99cc-4fcfa15d943b.png)

###########################################

![Screenshot from 2020-09-26 15-27-40](https://user-images.githubusercontent.com/6443210/94348840-ea0abd00-002e-11eb-8e9a-be97fc92a6f1.png)

PastaPastaPasta reviewed (approved):
ACK

xdustinface commented:
Two more fixes: 

c5390d5ea910b61eb5190f55f900a5b59152e454 To make sure all widgets receive the font weight updates if "font-weight-normal" gets updated.

95a169842b4e91a8704e8946c6e43de84a14e47c To make sure a widget which for example had a bold font initially gets updated to normal font if requested. Like all tab-like buttons we have, the currently clicked one is bold, the others normal. Without this commit they don't update properly if the font-weight gets adjusted in appearance tab. 

Those behaviours were broken by the changes of this PR, e.g. 2bfe078b2ff303486ebc47f412ad802243415f94 and 41524ea52e154667ed34364860868702e218d226

PastaPastaPasta reviewed (approved):
ACK

UdjinM6 commented:
Needs rebase

xdustinface commented:
> Needs rebase

Done

UdjinM6 reviewed (commented):
See below + (maybe) d734bf1 + found 2 issues:
1. overlay and status bar font issue: start the wallet (with --regtest to have the overlay popup shown), go to Options, change font size to lowest, close, start again (everything seems normal), go to Options and change font size to the largest. Everything else is huge but not overlay and status bar texts, it's like they have some medium size font now. Restart the wallet and now they are huge too. Same works the other way around: start with huge fonts, change to tiniest (medium size for overlay and status bar again), restart (now they are tiny too).

2. by tweaking PS visibility and width on start we override main window geometry settings. Should either respect it somehow _maybe_ ~or stop saving/loading it cause it doesn't make a lot of sense otherwise.~

EDIT: as discussed in Slack, n1 is an edge case so it's ok to fix it later and window geometry in n2 is still useful for window positioning

xdustinface commented on src/qt/bitcoingui.cpp:
```diff
@@ -1171,6 +1174,30 @@ void BitcoinGUI::updatePrivateSendVisibility()
+    // Reset the max width after it has been set to still allow window resizing
+    QTimer::singleShot(0, this, [=]() {
+        setMaximumWidth(std::numeric_limits<int>::max());
+    });
```
Jup thats better 👍

PastaPastaPasta reviewed (approved):
ACK

xdustinface commented:
> See below + (maybe) [d734bf1](https://github.com/dashpay/dash/commit/d734bf1bcb01f25d5109bb425749863a68603996) + found 2 issues:

In general im with you, having a smaller min size is preferable but i would rather revisit this sometimes later i tested it a bit and it breaks things with other font sizes and seems to mess a bit with traditional too.

The other commits i just added should fix the remaining scaling issues + they remove (as discussed in slack) the fixed pitch font. About the scaling: The previous default font implementation was not optimal because widgets without a object name set received an empty key from `getKey()`. The initial idea there was to only add the default font size for an instance of a class /UI element only once not for each instance. Now with d82c5d7 they are stored for the actual instance and removed again if the instance is not longer available in the UI.

PastaPastaPasta commented:
Nah, I think https://github.com/dashpay/dash/commit/cb23aae4204705f44fbe26ee983e0bb688f75303 is related enough. It's fine

PastaPastaPasta reviewed (approved):
ACK

UdjinM6 reviewed (approved):
👍 

ACK

strophy commented:
No, the problem with different size font in the menu and title bar remains. It is impossible to get them to match, and I think the reason is they are using different layout engines. The kerning looks really bad, letters have too much or not enough spacing on each side, icons in the menu are missing padding and it generally just doesn't feel "native" for Windows. I can count 6 different font sizes and weights in this screenshot alone, not counting the Dash logo...
![image](https://user-images.githubusercontent.com/32928115/94784333-7e5b9200-042a-11eb-9c3f-17dac3580f7b.png)

bitterseed commented:
![dash](https://user-images.githubusercontent.com/61271102/94790550-c7571a80-0419-11eb-924a-e42e5f3280d1.png)

Also having some scaling issues on ubuntu with a 1920x1080 display
Zooming in until my toolbar matches yours will give you a good approximation of what it looks like on my display

Cheers
